client/client.go

```diff
@@ -163,44 +170,231 @@ func makeURL(base string, parts []string) string {
 		c.setToken("")
-		err = c.sendAuthRequest(method, svcType, apiCall, requestData)
+		err = c.sendAuthRequest(method, svcType, apiVersion, apiCall, requestData)
 	}
 	return
 }
 
-func (c *authenticatingClient) sendAuthRequest(method, svcType, apiCall string, requestData *goosehttp.RequestData) (err error) {
+func (c *authenticatingClient) sendAuthRequest(
+	method, svcType, apiVersion, apiCall string,
+	requestData *goosehttp.RequestData,
+) (err error) {
 	if err = c.Authenticate(); err != nil {
 		return
 	}
 
-	url, err := c.MakeServiceURL(svcType, []string{apiCall})
+	url, err := c.MakeServiceURL(svcType, apiVersion, []string{apiCall})
 	if err != nil {
 		return
 	}
 	return c.sendRequest(method, url, c.Token(), requestData)
 }
 
-func (c *authenticatingClient) MakeServiceURL(serviceType string, parts []string) (string, error) {
+func (c *authenticatingClient) MakeServiceURL(serviceType, apiVersion string, parts []string) (string, error) {
 	if !c.IsAuthenticated() {
 		return "", errors.New("cannot get endpoint URL without being authenticated")
 	}
-	url, ok := c.serviceURLs[serviceType]
+	serviceURL, ok := c.serviceURLs[serviceType]
 	if !ok {
 		return "", errors.New("no endpoints known for service type: " + serviceType)
 	}
-	return makeURL(url, parts), nil
+	if serviceType == "object-store" {
+		// object-store has no versioning capability.
+		return makeURL(serviceURL, parts), nil
+	}
+	if apiVersion == "" {
+		// some tests don't include a version
+		return makeURL(serviceURL, parts), nil
+	}
+	if strings.Contains(serviceURL, c.serviceURLs["object-store"]) {
+		// a container endpoint won't have a version
+		return makeURL(serviceURL, parts), nil
+	}
+	requestedVersion, err := parseVersion(apiVersion)
+	if err != nil {
+		return "", err
+	}
+	rootURL, serviceURLSuffix, versions, err := c.getAPIVersions(serviceURL)
+	if err != nil {
+		return "", err
+	}
+	serviceURL, err = getAPIVersionURL(rootURL, serviceURLSuffix, versions, requestedVersion)
+	if err != nil {
+		return "", err
+	}
+	return makeURL(serviceURL, parts), nil
+}
+
+func getAPIVersionURL(rootURL *url.URL, serviceURLSuffix string, versions []apiVersionInfo, requested apiVersion) (string, error) {
+	var match string
+	for _, v := range versions {
+		if v.Version.major != requested.major {
+			continue
+		}
+		if requested.minor != -1 && v.Version.minor != requested.minor {
+			continue
+		}
+		for _, link := range v.Links {
+			if link.Rel != "self" {
+				continue
+			}
+			hrefURL, err := url.Parse(link.Href)
+			if err != nil {
+				return "", err
+			}
+			match = hrefURL.Path
+		}
+		if requested.minor != -1 {
+			break
+		}
+	}
+	if match == "" {
+		return "", fmt.Errorf("could not find matching URL")
+	}
+	versionURL := *rootURL
+	versionURL.Path = path.Join(versionURL.Path, match, serviceURLSuffix)
+	return versionURL.String(), nil
+}
+
+type apiVersion struct {
+	major int
+	minor int
+}
+
+func (v *apiVersion) UnmarshalJSON(b []byte) error {
```

Comment:
All this api version stuff probably is best placed in a separate file like apiversion.go